model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticatedClientSessionAdapter.java

```diff
@@ -40,56 +45,51 @@
+    private UserSessionModel userSession;
+
+    public AuthenticatedClientSessionAdapter(AuthenticatedClientSessionEntity entity, ClientModel client,
+                                             UserSessionModel userSession,
+                                             InfinispanChangelogBasedTransaction<String, UserSessionEntity> userSessionUpdateTx,
+                                             InfinispanChangelogBasedTransaction<UUID, AuthenticatedClientSessionEntity> clientSessionUpdateTx) {
+        if (userSession == null) {
+            throw new NullPointerException("userSession must not be null");
+        }
 
-    public AuthenticatedClientSessionAdapter(AuthenticatedClientSessionEntity entity, ClientModel client, UserSessionAdapter userSession,
-                                             InfinispanUserSessionProvider provider, InfinispanChangelogBasedTransaction updateTx) {
-        this.provider = provider;
         this.entity = entity;
-        this.client = client;
-        this.updateTx = updateTx;
         this.userSession = userSession;
+        this.client = client;
+        this.userSessionUpdateTx = userSessionUpdateTx;
+        this.clientSessionUpdateTx = clientSessionUpdateTx;
     }
 
     private void update(UserSessionUpdateTask task) {
-        updateTx.addTask(userSession.getId(), task);
+        userSessionUpdateTx.addTask(userSession.getId(), task);
     }
 
+    private void update(ClientSessionUpdateTask task) {
+        clientSessionUpdateTx.addTask(entity.getId(), task);
+    }
 
     @Override
-    public void setUserSession(UserSessionModel userSession) {
+    public void detachFromUserSession() {
         String clientUUID = client.getId();
 
-        // Dettach userSession
-        if (userSession == null) {
-            UserSessionUpdateTask task = new UserSessionUpdateTask() {
-
-                @Override
-                public void runUpdate(UserSessionEntity sessionEntity) {
-                    sessionEntity.getAuthenticatedClientSessions().remove(clientUUID);
-                }
-
-            };
-            update(task);
-            this.userSession = null;
-        } else {
-            this.userSession = (UserSessionAdapter) userSession;
-            UserSessionUpdateTask task = new UserSessionUpdateTask() {
-
-                @Override
-                public void runUpdate(UserSessionEntity sessionEntity) {
-                    AuthenticatedClientSessionEntity current = sessionEntity.getAuthenticatedClientSessions().putIfAbsent(clientUUID, entity);
-                    if (current != null) {
-                        // It may happen when 2 concurrent HTTP requests trying SSO login against same client
-                        entity = current;
-                    }
-                }
-
-            };
-            update(task);
-        }
+        UserSessionUpdateTask userSessionTask = new UserSessionUpdateTask() {
```

Comment:
If it's possible to avoid userSession update when detaching clientSession, it may be good. As we discussed on BJ. Leaving comment just so you not forget :)

Reply:
Good point, thanks. Fixed in the new commit.

Reply:
Note that it has some implications in handling the predicate so in the end I've taken a path that should be reasonably performant, trying to avoid writes to user session until a point when there are too many invalid client sessions in the user session accumulated, only then it gets updated.